fetch/fetch.go

```diff
@@ -0,0 +1,526 @@
+		return
+	}
+
+	f.activeBatchM.RLock()
+	batch, has := f.activeBatches[response.ID]
+	f.activeBatchM.RUnlock()
+	if !has {
+		f.log.Error("unknown batch response received, or already invalidated %v", response.ID)
+		return
+	}
+
+	// convert Requests to map so it can be invalidated when reading Responses
+	batchMap := batch.ToMap()
+	// iterate all hash Responses
+	for _, resID := range response.Responses {
+		//take lock here to make handling of a single hash atomic
+		f.activeReqM.Lock()
+		// for each hash, call its callbacks
+		resCallbacks := f.activeRequests[resID.Hash]
+		for _, req := range resCallbacks {
+			if req.validateResponse == false {
+				err := req.success(resID.Hash, resID.data)
+				req.returnChan <- HashDataPromiseResult{
+					Err:  err,
+					Hash: resID.Hash,
+				}
+				continue
+			}
+
+			// check hash flow, if hash didn't match still call fail function
+			actual := types.CalcHash32(data)
+			if actual == resID.Hash {
+				err := req.success(resID.Hash, data)
+				// if data is validates - put the returned value into the db.
+				if err == nil {
+					err = f.dbs[req.hint].Put(resID.Hash.Bytes(), data)
+				}
+				req.returnChan <- HashDataPromiseResult{
+					Err:  err,
+					Hash: resID.Hash,
+				}
+				continue
+			}
+			//todo: mark peer as malicious
+			err := fmt.Errorf("hash didnt match")
+			req.returnChan <- HashDataPromiseResult{
+				Err:  err,
+				Hash: resID.Hash,
+			}
+		}
+		//remove from map
+		delete(batchMap, resID.Hash)
+
+		// remove from active list
+		delete(f.activeRequests, resID.Hash)
+		f.activeReqM.Unlock()
+	}
+
+	//iterate all Requests that didn't return value from peer and invalidate them with error
+	err = fmt.Errorf("hash did not return")
+	for h := range batchMap {
+		f.activeReqM.Lock()
+		// for each hash, call its callbacks
+		resCallbacks := f.activeRequests[h]
+		f.activeReqM.Unlock()
+		for _, req := range resCallbacks {
+			req.returnChan <- HashDataPromiseResult{
+				Err:  err,
+				Hash: h,
+			}
+		}
+	}
+
+	//delete the hash of waiting batch
+	f.activeBatchM.Lock()
+	delete(f.activeBatches, response.ID)
+	f.activeBatchM.Unlock()
+}
+
+// this is the main function that sends the hash request to the peer
+func (f *Fetch) requestHashBatchFromPeers() {
+	var requestList []types.Hash32
+	f.activeReqM.RLock()
+	for hash := range f.activeRequests {
+		requestList = append(requestList, hash)
+	}
+	f.activeReqM.RUnlock()
+
+	// send in batches
+	for i := 0; i < len(requestList); i += f.cfg.BatchSize {
+		j := i + f.cfg.BatchSize
+		if j > len(requestList) {
+			j = len(requestList)
+		}
+		f.sendBatch(requestList[i:j])
+	}
+}
+
+func (f *Fetch) sendBatch(hashes []types.Hash32) { // create a network batch and serialise to bytes
+	// build list of batch messages
+	var batch requestBatch
+	f.activeReqM.RLock()
+	for _, hash := range hashes {
+		req, ok := f.activeRequests[hash]
+		if !ok {
+			f.log.Warning("message invalidated before sent in batch %v", hash)
```

Comment:
I think this case can happen if a malicious node tries to send unwanted data - we should be alerted of this

Reply:
If we should be warned, this is the right place for it, but rather when we process the unwanted data.  Ideally we should not allow a peer to invalidate requests which were not sent to him.

Here we're about to send requests which are only yet known locally. If a peer managed to invalidate a request before it was ever sent to anyone, it means it knew about its hash advance (in practice, if a peer managed to do so, it would be removed from `f.activeRequests` as well, so unless it happened after you aggregated the hashes in `requestHashBatchFromPeers` and before you entered here, it's won't affect this check).

So It looks to me just like a technical issue because you made `sendBatch` to accept `[]types.Hash32` instead `[]request` - both taken and re-taken from the same map. `request` already contains the hash - why not to use that instead? This is will also allow you to avoid adding high-priority requests to `f.activeRequests`.

Reply:
I've used requestMEssage since there could be multiple requests for a single hash

Reply:
How is this possible, If you're using a map with the hash as the key?

Reply:
there could be different callers that try to fetch same hash.